fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import Any, Tuple
+
+import torch
+from torch import nn
+import torch.distributed as dist
+import torch.nn.functional as F
+
+# Debugging flag to enable some prints. Useful to debug with FSDP.
+DEBUG = False
+
+
+def _next_power_of_2_or_max(n: int, max_n: int) -> int:
+    """ Return the smallest power of 2 greater than or equal to n, with a limit.
+
+        Useful when used in splitting a tensor into chunks with power-of-2 sizes.
+    """
+    # special case, just split to 1 element chunks.
+    if n == 0:
+        return 1
+    orig_n = n
+    n -= 1
+    n |= n >> 1
+    n |= n >> 2
+    n |= n >> 4
+    n |= n >> 8
+    n |= n >> 16
+    n += 1
+    assert n >= orig_n, f"{n} vs. {orig_n}"
+    assert bin(n).count("1") == 1, bin(n)  # Catch the case n is too large for this function.
+    if n > max_n:
+        return max_n
+    return n
+
+
+def _reshape_inputs(input: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Convert 3D inputs to 2D for this kernel"""
+    if len(input.shape) == 3:
+        input = input.reshape(-1, input.shape[2])
+    if len(target.shape) == 2:
+        target = target.reshape(-1)
+    return input, target
+
+
+def get_data(
+    shape: Tuple[Tuple[int, int], Tuple[int, int]], dtype: torch.dtype = torch.float16, device: str = "cuda"
+) -> Tuple[torch.Tensor, nn.Parameter, torch.Tensor]:
+    """ Utility function for getting some tensors for testing and benchmarking."""
+    (tokens, d1), (d2, vocabs) = shape
+    assert d1 == d2
+    input = torch.rand(tokens, d1, device=device, dtype=dtype).requires_grad_(True)
+    # Before pytorch 1.9, nn.Linear does not support device and dtype init option. So we use to()
+    # and an if condition.
+    layer = nn.Linear(d2, vocabs, bias=False).to(device)
+    assert dtype in [torch.float16, torch.float32]
+    if dtype == torch.float16:
+        layer = layer.half()
+    weight = layer.weight
+    target = (torch.rand(tokens, device=device) * vocabs).long()
+    return input, weight, target
+
+
+class BaselineSoftmax(nn.Module):
+    """ Baseline softmax that does an output linear projection and a softmax.
+
+        This is intended to be used with an embedding layer with shared weights.
+
+    Args:
+        proj_weight (nn.Parameter):
+            The shared weight.
+        tile_factor (int):
+            Unused. It is here to make kernel init easier with MEVO.
+        log (bool):
```

Comment:
nit: I would say "use_log" instead of "log". makes it more clear that it is a bool.